### PR TITLE
[Minor] Extract interval parsing logic, add unit tests

### PR DESCRIPTION
**Pull request by alamb**
# Which issue does this PR close?

N/A

 # Rationale for this change

Something I noticed while working on https://github.com/apache/arrow-datafusion/pull/2981/files was that the parsing of strings like `3 days` to interval constants was in the sql parser logic directly which made it somewhat hard to follow

Since https://github.com/sqlparser-rs/sqlparser-rs/pull/517  added support for arbitrary `Expr`s as interval arguments (rather than just strings), I thought refactoring this code into a function would be a step towards supporting generic arguments

It also made it it easier to write tests

I was thinking eventually we could use some of the code from @avantgardnerio from https://github.com/apache/arrow-rs/pull/2031  in Arrow but before I did that  I wanted to make sure the existing code had reasonable test coverage (to make sure we didn't break things).

# What changes are included in this PR?
1. Pull interval parsing logic into its own module and function. I tried to not make any logical changes
2. Add some (very) basic tests
3. Add functions to easily create `ScalarValue::Interval*`

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/sql/src/interval.rs [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+//! Interval parsing logic
+use datafusion_common::{DataFusionError, Result, ScalarValue};
+use std::str::FromStr;
+
+const SECONDS_PER_HOUR: f32 = 3_600_f32;
+const MILLIS_PER_SECOND: f32 = 1_000_f32;
+
+/// Parses a string with an interval like `'0.5 MONTH'` to an
+/// appropriately typed [`ScalarValue`]
+pub(crate) fn parse_interval(leading_field: &str, value: &str) -> Result<ScalarValue> {
```
Comment (alamb) [inline datafusion/sql/src/interval.rs:27]: This code was moved as closely as possible from parser.rs

File: datafusion/sql/src/planner.rs [hunk 2]
```diff
@@ -2216,146 +2217,12 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
...
-        // Month interval
-        if result_month != 0 {
-            return Ok(Expr::Literal(ScalarValue::IntervalYearMonth(Some(
-                result_month as i32,
-            ))));
-        }
+        let leading_field = leading_field
+            .as_ref()
+            .map(|dt| dt.to_string())
+            .unwrap_or_else(|| "second".to_string());
```
Comment (alamb) [inline datafusion/sql/src/planner.rs:2223]: This is the same code as used previously, but it has been raised up and passed in https://github.com/apache/arrow-datafusion/pull/2984/files#diff-c01a34949db6258aa1593f011ecf90f62cbde406acd5cdbf8b9b60b970ace1cfL2294-L2297

File: datafusion/sql/src/interval.rs [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+}
+
+#[cfg(test)]
+mod test {
+    use crate::assert_contains;
+
+    use super::*;
+
+    #[test]
+    fn test_parse_ym() {
```
Comment (alamb) [inline datafusion/sql/src/interval.rs:159]: These tests are new

File: datafusion/sql/src/interval.rs [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+        );
+
+        assert_eq!(
+            parse_interval("months", "7 days 5 minutes").unwrap(),
+            ScalarValue::new_interval_dt(7, (5.0 * 60.0 * MILLIS_PER_SECOND) as i32)
+        );
+    }
+
+    #[test]
+    // TODO file a ticket to track
```
Comment (alamb) [inline datafusion/sql/src/interval.rs:205]: I am not quite sure what to make of this -- my current interpretation is that `IntervalMonthDayNano` and the interval parsing logic are inconsistent but I wanted to double check

cc @avantgardnerio

**Review by alamb — COMMENTED** [review]
@ovr  I would appreciate a review if you had time

File: datafusion/sql/src/interval.rs (hunk 4 above)
Comment (avantgardnerio) [reply datafusion/sql/src/interval.rs:205]: When I was doing my implementation, `MonthDayNano` was the one I could never figure out how to test, because I could not get a SQL query to cause one to be generated. Clearly I was unaware of the `parse_interval()` function. I inferred the order of the values based on github conversations, documentation, and how they were ordered in `IntervalDayTimeType`.

So it's possible that `new_interval_mdn()` is not correct in its assumption:
```
        let m = months as u128 & u32::MAX as u128;
        let d = (days as u128 & u32::MAX as u128) << 32;
        let n = (nanos as u128) << 64;
```

I'll pull this branch down and see if I can add more clarity.

Comment (avantgardnerio) [reply datafusion/sql/src/interval.rs:205]: Looks like `parse_interval()` sees it as:
```
    if result_month != 0 && (result_days != 0 || result_millis != 0) {
        let result: i128 = ((result_month as i128) << 96)
                | ((result_days as i128) << 64)
                // IntervalMonthDayNano uses nanos, but IntervalDayTime uses milles
                | ((result_millis * 1_000_000_i64) as i128);
```

Comment (alamb) [reply datafusion/sql/src/interval.rs:205]: It is also entirely possible the but is in  `parse_interval` -- I am not sure anyone uses the `IntervalMonthDayNano` yet -- it was added relatively recently as I recall

Comment (avantgardnerio) [reply datafusion/sql/src/interval.rs:205]: Relevant conversation:
https://github.com/apache/arrow-datafusion/pull/2797/files#r910183035
https://github.com/apache/arrow-datafusion/pull/2797/files#r910186338

Comment (avantgardnerio) [reply datafusion/sql/src/interval.rs:205]: I think I found the C impl, and I think `parse_interval()` looks correct https://github.com/apache/arrow/blob/02c8598d264c839a5b5cf3109bfd406f3b8a6ba5/cpp/src/arrow/type.h#L1475

I can make a PR to fix shortly if we're sure this is the correct format. It would be good to have an integration test between C & Rust. Do we have anything like that now?

Comment (avantgardnerio) [reply datafusion/sql/src/interval.rs:205]: Linking to https://github.com/apache/arrow-rs/pull/2235 for resolution.

Comment (alamb) [reply datafusion/sql/src/interval.rs:205]:
```suggestion
    // https://github.com/apache/arrow-rs/pull/2235
```

**Review by yjshen — APPROVED** [review]

File: datafusion/common/src/scalar.rs [hunk 5]
```diff
@@ -618,6 +619,28 @@ impl ScalarValue {
             precision, scale
         )))
     }
+
+    /// Returns a [`ScalarValue::IntervalYearMonth`] representing
+    /// `years` years and `months` months
```
Comment (avantgardnerio) [inline datafusion/common/src/scalar.rs:624]: These seem very nice to have. I think more convenience functions like this are the key to terser and more readable code.

File: datafusion/sql/src/interval.rs (hunk 3 above)
Comment (avantgardnerio) [reply datafusion/sql/src/interval.rs:159]: And thank you for them! Comparing the results of differing implementations is an extremely useful test.

**Review by avantgardnerio — APPROVED** [review]
